sim/midas/src/main/cc/emul/verilator-harness.cc

```diff
@@ -1,141 +1,148 @@
-
-  top->ctrl_w_valid = m->w_valid();
-  top->ctrl_w_bits_strb = m->w_strb();
-  top->ctrl_w_bits_last = m->w_last();
-
-  top->ctrl_r_ready = m->r_ready();
-  top->ctrl_b_ready = m->b_ready();
-  memcpy(&top->ctrl_w_bits_data, m->w_data(), CTRL_BEAT_BYTES);
-
-  top->dma_aw_valid = d->aw_valid();
-  top->dma_aw_bits_id = d->aw_id();
-  top->dma_aw_bits_addr = d->aw_addr();
-  top->dma_aw_bits_size = d->aw_size();
-  top->dma_aw_bits_len = d->aw_len();
-
-  top->dma_ar_valid = d->ar_valid();
-  top->dma_ar_bits_id = d->ar_id();
-  top->dma_ar_bits_addr = d->ar_addr();
-  top->dma_ar_bits_size = d->ar_size();
-  top->dma_ar_bits_len = d->ar_len();
-
-  top->dma_w_valid = d->w_valid();
-  top->dma_w_bits_strb = d->w_strb();
-  top->dma_w_bits_last = d->w_last();
-
-  top->dma_r_ready = d->r_ready();
-  top->dma_b_ready = d->b_ready();
+  top->ctrl_aw_valid = simif_emul_t::simif_emul_t::master->aw_valid();
+  top->ctrl_aw_bits_id = simif_emul_t::master->aw_id();
+  top->ctrl_aw_bits_addr = simif_emul_t::master->aw_addr();
+  top->ctrl_aw_bits_size = simif_emul_t::master->aw_size();
+  top->ctrl_aw_bits_len = simif_emul_t::master->aw_len();
+
+  top->ctrl_ar_valid = simif_emul_t::master->ar_valid();
+  top->ctrl_ar_bits_id = simif_emul_t::master->ar_id();
+  top->ctrl_ar_bits_addr = simif_emul_t::master->ar_addr();
+  top->ctrl_ar_bits_size = simif_emul_t::master->ar_size();
+  top->ctrl_ar_bits_len = simif_emul_t::master->ar_len();
+
+  top->ctrl_w_valid = simif_emul_t::master->w_valid();
+  top->ctrl_w_bits_strb = simif_emul_t::master->w_strb();
+  top->ctrl_w_bits_last = simif_emul_t::master->w_last();
+
+  top->ctrl_r_ready = simif_emul_t::master->r_ready();
+  top->ctrl_b_ready = simif_emul_t::master->b_ready();
+  memcpy(
+      &top->ctrl_w_bits_data, simif_emul_t::master->w_data(), CTRL_BEAT_BYTES);
+
+  top->dma_aw_valid = simif_emul_t::dma->aw_valid();
+  top->dma_aw_bits_id = simif_emul_t::dma->aw_id();
+  top->dma_aw_bits_addr = simif_emul_t::dma->aw_addr();
+  top->dma_aw_bits_size = simif_emul_t::dma->aw_size();
+  top->dma_aw_bits_len = simif_emul_t::dma->aw_len();
+
+  top->dma_ar_valid = simif_emul_t::dma->ar_valid();
+  top->dma_ar_bits_id = simif_emul_t::dma->ar_id();
+  top->dma_ar_bits_addr = simif_emul_t::dma->ar_addr();
+  top->dma_ar_bits_size = simif_emul_t::dma->ar_size();
+  top->dma_ar_bits_len = simif_emul_t::dma->ar_len();
+
+  top->dma_w_valid = simif_emul_t::dma->w_valid();
+  top->dma_w_bits_strb = simif_emul_t::dma->w_strb();
+  top->dma_w_bits_last = simif_emul_t::dma->w_last();
+
+  top->dma_r_ready = simif_emul_t::dma->r_ready();
+  top->dma_b_ready = simif_emul_t::dma->b_ready();
 #if DMA_DATA_BITS > 64
-  memcpy(top->dma_w_bits_data, d->w_data(), DMA_BEAT_BYTES);
+  memcpy(top->dma_w_bits_data, simif_emul_t::dma->w_data(), DMA_BEAT_BYTES);
 #else
-  memcpy(&top->dma_w_bits_data, d->w_data(), DMA_BEAT_BYTES);
+  memcpy(&top->dma_w_bits_data, simif_emul_t::dma->w_data(), DMA_BEAT_BYTES);
 #endif
```

Comment:
@davidbiancolin explained offline that both VCS and Verilator have a bitwidth where the underlying type changes and the memcpy calls change from pass by value to pass by reference.

We both agree that it would be nice to avoid this with some abstraction layer in the rewrite, possibly DPI.